### PR TITLE
Applications: Cytoland refactor

### PR DESCRIPTION
**Pull request by alxndrkalinin**
## Summary

Refactors the virtual staining translation module into a standalone `viscy-translation` application package, syncs with `app-dynaclr` (new DynaCLR features, data infrastructure, evaluation framework), and ports PR #280 sliding window prediction.

## Translation application (Phase 26)

Extracted `viscy/translation/` into `applications/translation/` (`viscy-translation` package):

- **Engine**: `VSUNet`, `FcmaeUNet`, `AugmentedPredictionVSUNet` LightningModules
- **Shared components** extracted to viscy-utils: `HCSPredictionWriter`, `MixedLoss`
- **Evaluation**: `SegmentationMetrics2D` stays app-specific
- **Lightning CLI**: `python -m viscy_translation fit --config fit.yml`
- **21 tests**: engine smoke tests, training integration (6 tests: VSUNet/FcmaeUNet with synthetic + real data), config class_path validation, inference reproducibility (HPC-gated), sliding window prediction

## PR #280 port: sliding window volume prediction

- Enhanced `_blend_in` in `prediction_writer.py` to support both `torch.Tensor` (5D) and `np.ndarray` (4D)
- Added `predict_sliding_windows()` to `AugmentedPredictionVSUNet` for in-memory Z-sliding inference
- Extracted `_predict_with_tta()` helper; made transforms optional (default identity)
- Added `getattr` guard for `out_stack_depth` (clear error for 2D models)

## app-dynaclr sync (PR #381 + #384)

Merged 108 commits from `app-dynaclr`:
- Multi-experiment training infrastructure (ExperimentRegistry, MultiExperimentIndex, MultiExperimentDataModule)
- New viscy-data modules: cell_index, collection, sampler, schemas, channel_dropout
- Foundation models: DINOv3, OpenPhenom; contrastive loss: NTXentHCL
- DynaCLR evaluation framework: linear classifiers, dimensionality reduction, pseudotime analysis
- New applications: airtable, qc
- Code review fixes: assert→ValueError, `_components/` deleted, organelle→marker rename, ArrowString compat

## Bug fixes

- Fix operator precedence in `evaluation.py`
- Fix source overwrite in `AugmentedPredictionVSUNet` TTA loop
- Fix unbound `return_target` in `FcmaeUNet.forward_fit_task`
- Add `weights_only=True` to `torch.load`; fix mutable default `model_config`
- Fix numpy indexing in `_blend_in`; respect `overwrite` flag in `EmbeddingWriter`
- Port `UNeXt2Decoder` Sequence mutation fix to `components/blocks.py`
- Port `UNeXt2Stem` ZeroDivisionError guards to `components/stems.py`
- Validate `every_n_epochs >= 1` in `EmbeddingSnapshotCallback`
- Fix CPU/GPU `pos_weight` device mismatch in `ClassificationModule`
- Fix `None` norm_meta crashing `default_collate` in `SlidingWindowDataset`
- Apply `torch.sigmoid` before `binary_accuracy`/`binary_f1_score` (logits→probs)
- Rename `ULTRACK_INDEX_COLUMNS` → `INDEX_COLUMNS`; standardize `select.py` imports

## Test results

| Suite | Result |
|-------|--------|
| viscy-models | 95 passed, 1 xfailed |
| viscy-data | 169 passed, 2 skipped |
| translation | 21 passed |
| dynaclr | 141 passed, 4 skipped |

See [deferred items](https://github.com/mehta-lab/VisCy/pull/379#issuecomment-4078859865) for tracked improvements.

### CODE REVIEW

File: packages/viscy-models/src/viscy_models/_components/blocks.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                conv_blocks=conv_blocks,
+                norm_name=norm_name,
+                upsample_pre_conv=upsample_pre_conv,
+            )
+            self.decoder_stages.append(stage)
+
+    def forward(self, features: Sequence[Tensor]) -> Tensor:
+        feat = features[0]
+        # padding
+        features.append(None)
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply packages/viscy-models/src/viscy_models/_components/blocks.py]: Fixed in c63d6d4. Replaced `features.append(None)` with `skips = list(features[1:]) + [None]` to avoid mutating the input Sequence.

File: packages/viscy-utils/src/viscy_utils/evaluation/annotation.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    >>> embeddings_ds = xr.open_zarr(embeddings_path)
+    >>> adata = convert(embeddings_ds, output_path, overwrite=True, return_anndata=True)
+    >>> adata
+    AnnData object with n_obs × n_vars = 18861 × 768
+        obs: 'id', 'fov_name', 'track_id', 'parent_track_id', 'parent_id', 't', 'y', 'x'
+        obsm: 'X_projections', 'X_pca', 'X_umap', 'X_phate'
+    """
+    # Check if output_path exists
+    if output_path.exists() and not overwrite:
+        raise FileExistsError(f"Output path {output_path} already exists.")
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply packages/viscy-utils/src/viscy_utils/evaluation/annotation.py]: Not a bug. AnnData's write_zarr() handles overwriting internally -- it clears the zarr store before writing. Verified by testing overwrite with different obs columns and obsm entries; no stale data persists.

File: applications/dynaclr/src/dynaclr/classification.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        Loss function, by default BCEWithLogitsLoss.
+    example_input_array_shape : tuple, optional
+        Shape of example input array.
+    """
+
+    def __init__(
+        self,
+        encoder: ContrastiveEncoder,
+        lr: float | None,
+        loss: nn.Module | None = nn.BCEWithLogitsLoss(pos_weight=torch.tensor(1.0)),
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply applications/dynaclr/src/dynaclr/classification.py]: Already fixed in commit 2be23de. BCEWithLogitsLoss registers pos_weight as a buffer, so .to(device) moves it automatically.

File: packages/viscy-utils/src/viscy_utils/callbacks/embedding_snapshot.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        features_np = self._features.numpy()
+        projections_np = self._projections.numpy()
+
+        if self._index is not None:
+            available = {k: v for k, v in self._index.items() if k in ULTRACK_INDEX_COLUMNS}
+            index_df = pd.DataFrame(available)
+        else:
+            index_df = pd.DataFrame({"fov_name": ["unknown"] * features_np.shape[0]})
+
+        uns_metadata = EmbeddingWriter._collect_data_provenance(trainer)
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply packages/viscy-utils/src/viscy_utils/callbacks/embedding_snapshot.py]: Acknowledged in earlier pass. Tracked for future refactor.

File: packages/viscy-utils/src/viscy_utils/callbacks/embedding_snapshot.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+        if self.store_images and self._images is not None:
+            import anndata as ad
+
+            adata = ad.read_zarr(output_path)
+            b = self._images.shape[0]
+            adata.obsm["X_images"] = self._images.reshape(b, -1)
+            # Mid-Z extraction produces (C, Y, X) per sample
+            adata.uns["image_shape_cyx"] = list(self._images.shape[1:])
+            adata.write_zarr(output_path)
```
[the remark this answers is not in the document]
Comment (alxndrkalinin) [reply packages/viscy-utils/src/viscy_utils/callbacks/embedding_snapshot.py]: Good optimization suggestion. Will consider extending write_embedding_dataset to accept optional obsm/uns payloads in a follow-up.

**Comment by alxndrkalinin** [comment]
### Deferred items from code review

The following issues were identified during Copilot and Claude code reviews. All actionable bugs have been resolved in recent merges (PR #381, #384, PR #280 port). These remaining items are acknowledged as out-of-scope improvements to address in future work.

**Resolved since last review:**
- `_components/` dead code -- directory deleted in PR #384; all fixes ported to `components/`
- `UNeXt2Decoder.forward` mutation -- fixed in `components/blocks.py`
- `UNeXt2Stem` ZeroDivisionError -- validation guards ported to `components/stems.py`

**Architecture / dependency hygiene:**
- `embedding_snapshot.py` imports `viscy_data` at runtime but `viscy-utils` does not declare `viscy-data` as a dependency. Works because applications depend on both, but architecturally incorrect.
- `embedding_snapshot.py` uses `EmbeddingWriter._collect_data_provenance` (private API). Should be extracted to a public utility function.
- `embedding_snapshot.py` re-opens zarr store after writing for obsm/uns updates. Could be optimized by extending `write_embedding_dataset()` to accept optional obsm/uns payloads.

**Minor code quality:**
- `classification.py:102` uses deprecated `nn.functional.sigmoid` in `predict_step`. Should use `torch.sigmoid`.
- `qc_metrics.py` uses `print` instead of `logging`. Should switch when CLI interface is finalized.
- `annotation.py` `convert()` has no unit tests for overwrite semantics or obsm construction.
- `annotation.py` uses `np.column_stack` on xarray coords without explicit `.data` conversion (works but fragile).

**Design choices (not bugs):**
- `qc_metrics.py` hardcodes focus-specific naming (`all_focus_values`, `z_focus_*`). Acceptable while QC app is focus-only; revisit when more metric types are added.
- `test_engine.py` depends on external `grep` executable. Acceptable for Linux-only HPC project.
- `cli.py` `isdisjoint` logic was flagged as inverted -- verified correct.
- `annotation.py` reads projections as coords, features as data_vars -- by design (EmbeddingWriter stores 1D projections as coords).
